Add Kemeny's constant

peijenburg opened the pull request:
Implemented Kemeny's constant for undirected and directed graphs.
The concept is borrowed from Markov theory. Viewing the graph as a Markov chain, Kemeny’s constant is the expected number of time steps required for a Markov chain to transition from a starting state i to a random destination state sampled from the Markov chain's stationary distribution. Surprisingly, this quantity does not depend on which starting state i is chosen.

The concept is more and more used in graph theory and network science, see e.g.
https://dl.acm.org/doi/abs/10.1145/3366423.3380093
https://epubs.siam.org/doi/abs/10.1137/22M1486728
https://www.sciencedirect.com/science/article/pii/S0024379517305256

dschult commented:
Could this go into the `distance_measures.py` module in algorithms?
Is that what this is used for?
Maybe a sentence or two in the doc_string about what this measures would be helpful too (right now it says how to compute it, but not what it is trying to measure).

Thanks!

peijenburg commented:
@dschult Thanks for your suggestions!

Although the Kemeny's constant isn't a distance measure in the strict sense (it doesn't necessarily decrease after adding more edges to the graph), it is often used to characteristics the connectivity of graphs.

I've added two lines in the documentation to explain this and also moved it to `distance_measures.py`

It should be ready for review now.

peijenburg commented:
I cannot figure out why one of the tests fails. It seems that `nx.is_negatively_weighted(G, weight="weight")` doesn't recognise that the graph has negative weights... Can someone check this?

dschult commented:
This test failure is very strange.  It only happens on the Ubuntu system with Python 3.11.  The runs with other systems work with other versions of Python on that system work. And the test just has G with 3 edges with integer weights. Nothing fancy... check if the `nx.is_negatively_weighted` function works. 

BTW the check for negative weights hardcodes the edge attribute named "weight" while the rest of the code uses an input parameter `weight`. You should probably have a test for the case of edge attributes named something else.  But that doesn't help figure out why this is working on Ubuntu and working on Python 3.11 but not working on Ubuntu with Python 3.11.

peijenburg commented:
@dschult I'm unsure what happened, but I made the minor changes you suggested and the error suddenly disappears...

MridulS commented:
> But that doesn't help figure out why this is working on Ubuntu and working on Python 3.11 but not working on Ubuntu with Python 3.11.

The failing test is coming from the dispatching test case, not the usual pytest case.

Before the last commit, the function was called with just `nx.kemeny_constant(G)` in the test suite and the current setup will only copy over the "weight" attributes to the `LoopbackGraph` if the weight attribute is specifically passed in the function call `nx.kemeny_constant(G, weight='weight')`. The error was fixed because the function signature was changed in the last commit.

One way to make sure that all edge attributes are preserved is to use  `preserve_edge_attrs` with the dispatch decorator.
``@nx._dispatch(edge_attrs="weight", preserve_edge_attrs=True)`` but we shouldn't really need to.

But anyway I think the dispatching test was able to find this (yay!) even though it's super straight forward to debug. I'm not sure who exactly we should be making it easier for new contributors to understand this bit 😅

dschult commented:
Thanks @MridulS !!  
And, yes, we should think about how to make it easier for new contributors to figure it out.

But why did only the Ubuntu test with Python 3.11 fail? 
If that test handles the loopback differently, maybe we can name the test somehow to indicate that it is testing the loopback feature. ??

dschult commented on networkx/algorithms/distance_measures.py:
```diff
@@ -769,3 +770,104 @@ def resistance_distance(G, nodeA, nodeB, weight=None, invert_weight=True):
+    Raises
+    ------
+    NetworkXError
+        If the graph `G` is directed.
+
+    NetworkXError
+        If the graph `G` is not connected.
+
+    NetworkXError
+        If the graph `G` does not contain any nodes.
+
+    NetworkXError
+        If the graph `G` contains edges with negative weights.
```
Let's combine these NetworkXErrors perhaps with description (change to improve as you like) 
```
        If the graph `G` is directed, or not connected,
        or contains no nodes, or has edges with negative weights.
```

dschult commented on networkx/algorithms/distance_measures.py:
```diff
@@ -769,3 +770,104 @@ def resistance_distance(G, nodeA, nodeB, weight=None, invert_weight=True):
+
+    Notes
+    -----
+    The implementation is based on equation (3.3) in [2]_. Self-loops are included.
+    Multi-edges are contracted in one edge with weight equal to the sum of the weights.
```
What does "self-loops included" refer to? 
Perhaps wording like:
"Self-loops are allowed and indicate a Markov Chain where
the state can remain the same. "

dschult commented on networkx/algorithms/distance_measures.py:
```diff
@@ -769,3 +770,104 @@ def resistance_distance(G, nodeA, nodeB, weight=None, invert_weight=True):
+    H = DH @ (A @ DH)
+
+    # Compute eigenvalues of H
+    eig = sp.linalg.eigvalsh(H.todense())
+    eig = sorted(eig)
```
I don't think the eigenvalues need to be sorted (based on the rest of the code).

And, its a shame to have to convert `H.todense()` because that could be a memory issue. But I don't see a way around it with the sparse eigen-solvers offered by Scipy in `sp.sparse.linalg`. They don't allow computation of all the eigenvalues.

peijenburg replied:
The eigenvalues must be sorted, because the largest eigenvalue (which equals 1) is not used in the formula for Kemeny's constant. You can also see why: then the formula explodes (and gives +infinity as a result)

For the computation of all eigenvalues, a dense matrix is generally preferred, as far as I know. Anyway, the computation of all eigenvalues is O(N^3) operation, making stuff slow for N >= 4,000.

dschult commented on networkx/algorithms/distance_measures.py:
```diff
@@ -769,3 +770,104 @@ def resistance_distance(G, nodeA, nodeB, weight=None, invert_weight=True):
+    if len(G) == 0:
+        msg = "Graph G must contain at least one node."
+        raise nx.NetworkXError(msg)
+    elif nx.is_directed(G):
```
These `elif` should just be `if` (because the previous `if` will always `raise` and thus stop the program flow.

Also, many of these error messages should be included in the raise call (removing the variable `msg`, but more importantly putting it all on one line -- that gets shown by the error traceback)

peijenburg replied:
First part, done.

Second part, I am not completely aware of your question, but I presume it amounts to changing
`msg = "Graph G must contain at least one node.";
raise nx.NetworkXError(msg)`

to

`raise nx.NetworkXError("Graph G must contain at least one node.")`

which I did

dschult commented on networkx/algorithms/tests/test_distance_measures.py:
```diff
@@ -472,3 +472,148 @@ def test_this_one_specific_tree(self):
+    def test_kemeny_constant_not_connected(self):
+        self.G.add_node(5)
+        with pytest.raises(nx.NetworkXError):
+            nx.kemeny_constant(self.G)
```
Add tests for all the NetworkXError cases (perhaps in one test function)

dschult reviewed (commented):
A few comments below (getting into the good stuff now). 
- In the papers, I see both `Kemeny's constant` and `the Kemeny constant` to describe this thing. The function name is what I would prefer -- `kemeny_constant` because it makes it a two-word-noun instead of an adjective and a noun. Also it is easier to type and (at least for me) to think about. (It also avoids the apostrophe which can mess with python's strings if you get unlucky.) If we are going that route, I would suggest the docs also use that terminology (perhaps with a phrase noting that it is also called "Kemeny's constant")
- Does this work for directed as well as undirected graphs?  If not use the decorator `@nx.utils.not_implemented_for("directed")` just before the dispatch decorator. Also you should ad a test with a DiGraph and check that it either works or raises an error.

A few tweaks to match our systems:
- since this is a new function, we should make the optional keyword required: `def kemeny_constant(G, *, weight=None):`  That is supposed to make it more clear that the second input value is the weight.
- please add this function to the "needs_numpy" and "needs_scipy" lists in `networkx/conftest.py` (~line 200)
- add a hook to this function in the docs in `doc/reference/algorithms/distance_measures.rst`.  You can check that this worked by clicking on the "Details" link of the CI-test called "document-artifact". that brings you to the documentation as it will appear after this PR. search for your function and check that it is rendering nicely.

More below.  Thanks for this!

peijenburg commented:
@dschult Thanks so much for the review!

> I would suggest the docs also use that terminology (perhaps with a phrase noting that it is also called "Kemeny's constant")
Done

> Does this work for directed as well as undirected graphs?
It only works for undirected graphs. Before, I had a NetworkXError thrown for directed graphs, but it seems networkx's standard is to use not_implemented_for(...), which sounds a little strange to me, because it is not just not implemented, it actually doesn't exist. But that's just terminology, I guess. Anyway, I changed it as you requested and added a test.

> make the optional keyword required
Done

> please add this function to the "needs_numpy" and "needs_scipy" lists
Done

> add a hook to this function in the docs
As far as I am aware, I already implemented that. Perhaps you didn't see it because the list is ordered alphabetically?

dschult reviewed (approved):
This Looks Good.   Thanks!

rossbar commented on networkx/algorithms/distance_measures.py:
```diff
@@ -769,3 +770,99 @@ def resistance_distance(G, nodeA, nodeB, weight=None, invert_weight=True):
+    DH = sp.sparse.csr_array(sp.sparse.spdiags(diags_sqrt, 0, m, n, format="csr"))
+    H = DH @ (A @ DH)
+
+    # Compute eigenvalues of H
+    eig = sp.linalg.eigvalsh(H.todense())
+    eig = sorted(eig)
+
+    # Compute the Kemeny constant
+    K = 0
+    for i in range(n - 1):
+        K += 1 / (1 - eig[i])
+    return K
```
Just a nit - since these are already arrays, might as well vectorize this which is easier to read IMO:

```suggestion
    # Compute eigenvalues of H
    eig = np.sort(sp.linalg.eigvalsh(H.todense()))

    # Compute the Kemeny constant
    return np.sum(1 / (1 - eig[:-1]))
```

rossbar commented on networkx/conftest.py:
```diff
@@ -210,6 +210,7 @@ def add_nx(doctest_namespace):
     "linalg/laplacianmatrix.py",
     "utils/misc.py",
     "algorithms/centrality/laplacian.py",
+    "algorithms/distance_measures/kemeny_constant.py",
```
This looks like it might be left over from an earlier version of this PR? AFAICT there is no file by this name.

Having noticed this, I was surprised that the CI was passing but it turns out we don't run doctests without default dependencies. I don't remember if this is intentional or not :upside_down_face: .

Either way - I propose to just remove these two added lines. Let's leave worrying about testing configurations to a different PR.
